File: 이진검색(이진탐색)/leetcode2187(Medium)-Minimum Time to Complete Trips.js
// 파라메트릭 서치 : 327ms(24.66%), 58.1MB(16.44%)
/**
 * @param {number[]} time
 * @param {number} totalTrips
 * @return {number}
 */
var minimumTime = function (time, totalTrips) {
  time = time.sort((a, b) => a - b)

  let left = time[0]
  let right = time[time.length - 1] * totalTrips

  const isEnough = (mid) => {
    let acc = 0
    for (const t of time) {
      acc += Math.floor(mid / t)
      if (acc >= totalTrips) {
        return true
      }
    }
    return false
  }

  while (left < right) {
    const mid = Math.floor(left / 2 + right / 2)
    if (isEnough(mid)) {
      right = mid
    } else {
      left = mid + 1
    }
  }

  return left
}

File: 이진검색(이진탐색)/leetcode2187(Medium)-Minimum Time to Complete Trips.py
# 파라메트릭 서치 : 2146ms(55.94%), 28.8MB(6.68%)
class Solution:
    def minimumTime(self, time: List[int], totalTrips: int) -> int:
        time.sort()
        left = time[0]
        right = time[-1] * totalTrips

        def is_enough(mid):
            acc = 0
            for t in time:
                acc += (mid // t)
                if acc >= totalTrips:
                    return True
            return False
            

        while left < right:
            mid = int((right / 2) + (left / 2)) # avoid overflow
            if is_enough(mid):
                right = mid
            else:
                left = mid + 1

        return left